### PR TITLE
feat(tags): RFC for tags

### PR DESCRIPTION
**Pull request by frsann**
## Checklist
- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: docs/rfc/active/tags/README.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+We want to introduce some new under `datahub/metadata-models/src/main/pegasus/com/linkedin/common/`.
+
+### `Tag` entity
+
+First we create a `TagMetadata` entity, which defines the actual tag-object.
+
+The URN of the tag consists for two parts: a category and a name. This gives us more flexibility as to how tags are use
+and reduces the risk of name collisions. An example of an `category` could be `sensitivity` and the name could be `pii`
+or `public`.
```
Comment (gabe-lyons) [inline docs/rfc/active/tags/README.md:61]: I was also thinking about adding a tag `category` field. Right now I'm favoring allowing tags on tags, and using the tags applied to a given tag as its "category". This way we won't have to create two different grouping systems.

File: docs/rfc/active/tags/README.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+```
+/**
+ * Tag information
+ */
+record TagEmployment {
+
+  /**
+   * Tag in question
+   */
+  tag: TagMetadata
```
Comment (gabe-lyons) [inline docs/rfc/active/tags/README.md:93]: I think we can get away with tag being just of type `TagUrn`

File: docs/rfc/active/tags/README.md [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+  /**
+   * Who has edit rights to this employment. WIP
+   */
+  edit: union[None, any, role-urn]
```
Comment (gabe-lyons) [inline docs/rfc/active/tags/README.md:98]: I like leaving this in the proposal, but lets keep an * here until we land on an access-control system for datahub as a whole. I think that will have to be a second parallel RFC, but hopefully we can push access control and tags forward together.

File: docs/rfc/active/tags/README.md [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+- RFC PR: (after opening the RFC PR, update this with a link to it and update the file name)
+- Discussion Issue: (GitHub issue this was discussed in before the RFC, if any)
+- Implementation PR(s): (leave this empty)
+
+# Tags
+
+## Summary
+
+Adding general support for tag lowers the bar to add simple metadata to entities. Tags can be used as part of the
+business logic (e.g. sensitivity tag on a data set) or more ad hoc use cases, such as a data analyst's favorite
```
Comment (jplaisted) [inline docs/rfc/active/tags/README.md:11]: I wonder if this is too generic of a goal? Especially these two examples.

For business logic tags, I would think people would want to action on that, especially via an API. We let people tag fields (with a well known set of tags), perhaps we should derive it from the fields. As far as PII/non-PII goes. If there's some other set of domain lexicon here (i.e. can be shared outside the company/internal only), that should probably be well known, rather than a free form tag? Especially if users are inputting it, a drop down would be better than a text field liable to typos.

For a data analyst's favorite, perhaps DataHub should just have a proper page for that? :) Or for "custom lists" or "view"? Just adding a tag like `plaisted-favorite` and then searching for said tag means that the view would just be DataHub's search page, rather than some nicer page. This one is more minor. Internally at LinkedIn, we have the concept of "Liked" and "Followed" datasets, and we do show them on your profile page, so they're easy to find. I don't think these "social features" exist in open source, but it something we could consider here.

I get tags are useful for adding more search terms, maybe we should sort of limit it to that? So for now the "favorite" use case is probably ok, but in the future we may have a better page. But I would think the primary use case is to add more search terms? Anything that is more specific than that, and especially if requires action, should be more purpose built (i.e. can be shared or not).

File: docs/rfc/active/tags/README.md [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+>
+> This list can be high level and not detailed. It is to help focus the conversation on what you want to focus on.
+
+## Detailed design
+
+We want to introduce some new under `datahub/metadata-models/src/main/pegasus/com/linkedin/common/`.
+
+### `Tag` entity
+
+First we create a `TagMetadata` entity, which defines the actual tag-object.
```
[the remark this answers is not in the document]
Comment (jplaisted) [reply docs/rfc/active/tags/README.md:57]: I'm assuming there should be some `value: string` field somewhere :)

**Comment by shirshanka** [comment]
Rendered: https://github.com/frsann/datahub/blob/tags-rfc/docs/rfc/active/tags/README.md

File: docs/rfc/active/tags/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (frsann) [reply docs/rfc/active/tags/README.md:57]: I assume you're referring to string value of the tag that, for example, would be shown in the UI?  So the `<name>` in the Urn is not suitable to be parsed for this purpose?

File: docs/rfc/active/tags/README.md (hunk 1 above)
Comment (frsann) [reply docs/rfc/active/tags/README.md]: Fixed as proposed.

File: docs/rfc/active/tags/README.md (hunk 2 above)
Comment (frsann) [reply docs/rfc/active/tags/README.md:93]: For sure! Fixed.

File: docs/rfc/active/tags/README.md (hunk 3 above)
Comment (frsann) [reply docs/rfc/active/tags/README.md:98]: Add a comment on this!

File: docs/rfc/active/tags/README.md (hunk 4 above)
Comment (frsann) [reply docs/rfc/active/tags/README.md:11]: You raise some good points here! I'll clean up the RFC today, and when it's in a bit more respectable form we can open up the discussion to the community. 

What I think we're aiming for is a very flexible and low threshold way of tagging entities, and it's up to each and every org to decide if and when and how to use it. I guess it can in some sense be reduced to "adding more search terms", and the examples I present are maybe the best ones (for the reason you presented), but at the same time it think there is value in adding a all-encompassing tagging system as it's the teams' imagination that sets the limit on how it can be utilised. 

I agree that some cases certainly benefits from a more purpose built solution (I want a sensitivity tag propagation feature 😉 ), but both of the cases mentioned above does feel like special cases of a generic tagging system.

Comment (frsann) [reply docs/rfc/active/tags/README.md]: I updated the RFC a bit, hope it clears up our thinking a bit more.

File: docs/rfc/active/tags/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (jplaisted) [reply docs/rfc/active/tags/README.md:54]: So conceptually I think there's a few ways to do this. I don't this proposal fits either right now, but I could be misreading.

1. Treat tags as first class entities (meaning they have an URN that uniquely identifies each tag). If I make the tag `plaisted-favorite` only one "copy" of it exists. Then add some "lineage" or other graph edge between the tag and tagged entities. I don't fully see the value in this, as this A) requires a graph search, which seems overkill, and B) I don't think we'll have any extra information on the tag, besides the string itself? Note that in this case, its probably still a good idea to normalize the data and have a string field for the tag name, even if the tag name is also in the URN. Separates data from identifier :) 
2. Treat tags as aspects of existing entities. In this case they don't have an URN, they aren't entities themselves. For an example of something like this today, look at `Status`. The advantage here is that tags are just part of entities (via aspects), and can be treated that way, meaning you can use the search index here rather than the graph index. If the idea is just to get more search terms for an entity, you probably want things easily accessible in the search index, not the graph index.


Consider something like this:

```
// Not sure if we need Tag or if an array[string] is ok in Tags. This is more future proof; allows to add more
// information a tag if we want easily...
record Tag {
  value: string;
}

/**
 * Aspect to tag entities with custom search terms.
 */
record Tags {
  tags: array[Tag];
}

// Now if we want to be able to tag datasets, lets add it to the list of dataset aspects
typeref DatasetAspect = union[
   // bunch of existing stuff...
   Tags
]

// And if we want it in the search index, lets add it to the search model
record DatasetDocument includes BaseDocument {
   // bunch of existing stuff...
   tags: array[string];
}
```

Granted, I guess downsides to this second approach is that you need to edit every entity definition you want to be able to tag, where as in the first approach you don't need to.


Maybe another way to phrase this is "are tags valuable nodes in the metadata graph, or are they just more pieces of metadata on existing entities?" Maybe flesh out the use cases more (again, total ack this is a WIP), and that question can be more easily answered.

[the remark this answers is not in the document]
Comment (jplaisted) [reply docs/rfc/active/tags/README.md:54]: Like, lets focus on the DataHuber user story first, and then we can figure out the implementation details :)

File: docs/rfc/active/tags/README.md (hunk 4 above)
Comment (jplaisted) [reply docs/rfc/active/tags/README.md]: Thanks for the update. I do think my "is this too generic" concern applies. Isn't all metadata just a tag? :p I'm hesitant to encourage users to use tags like this, because it can lead to overuse/abuse of unstructured, manually entered data. Again, I'm fine with it if the idea is just adding some more search terms. But I think once a tag becomes "actionable" or needs a fixed lexicon, then it should become first class data.

I believe part of the idea behind your proposal to allow tagging tags was to group tags. To me, this is you trying to have that standard lexicon. e.g. imagine the exporting data to other companies case. Maybe I make the tags "internal" and "shareable", and then tag both tags with "3rd-party-tags". My issue here is that this is liable to be typo'd, under documented, hard to discover (at least this proposal doesn't talk about being able to document and browse for tags in a UI), and lacks governance (can I make the tag "cheese" and then tag it with "3rd-party-tags" with no approval?). In fact, for this case, whether something is allowed to be shared with a third party, probably isn't a good case for manual tagging anyway, even with an agreed upon lexicon; you should probably be deriving that information via PII tags?

I think once something crosses some vaguely grey line (more than just "search terms", anything you'd want to action upon), you want some extra guard rails in place, good documentation, and governance. To me, this means just modeling out this space as actual metadata! Just like we do with field sensitivity. Granted, I saw _we_, but it looks like it isn't open sourced, so I guess I mean Linkedin internally.

This is a super simplified version of what we have for field compliance internally:

```pdl
enum ComplianceDataType {
   /**
   * Personal physical address
   */
  ADDRESS

  /**
   * City, State, Region, etc
   */
  CITY_STATE_REGION

  NOT_PII

  // etc, other well known PII types goes here, like emails, names, etc
}

record FieldCompliance {
  // path to this field from the root of the schema
  fieldPath: string
  
  /**
   * The business / semantic meaning or data type of data fields
   */
  dataType: ComplianceDataType
}

record ComplianceInfo {
  complianceFields: array[FieldCompliance]
}

typeref DatasetAspect = union[
  // etc
  ComplianceInfo
]
```

So rather than do this all via tags, we've done it in the actual metadata ecosystem, to make it well defined. In our UI, you can change the PII tags of fields, but its via a drop down that shows these values.


So, again, lets maybe nail down the user story here. When would you use a tag vs editing the actual metadata models? When would you tag tags? To me, at least at the surface, a tag is just adding more terms into the search index. ideally nothing actions on that, I'm just trying to make it easier for other humans to find. If anything wants to action on that, especially for legal processes, it should be well defined, documented, and governed :)

File: docs/rfc/active/tags/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (frsann) [reply docs/rfc/active/tags/README.md:54]: Got it! We're definitely going for the use case 1 ☝️. I tried to make the case for this approach in the RFC, so it adds little value to repeat that explanation here. Thanks, for the clarification, I will add the `name` field to the `TagMetadata`.

File: docs/rfc/active/tags/README.md (hunk 4 above)
Comment (frsann) [reply docs/rfc/active/tags/README.md]: > this is liable to be typo'd 
This is a UI problem in my opinion. As I explain in the "Alternatives" section, how Apache Atlas does this is that it has a separate flow for creating tags, and the attachment is done using a dropdown list. This reduces the risk of typos.  

> under documented
I added a description and a external description to the `TagMetadata`. Do you think this helps?

> hard to discover (at least this proposal doesn't talk about being able to document and browse for tags in a UI), 
Yes, I ruled it out in the "Non-Requirements" section, because I think it would explode the scope of the RFC. Again, I think Atlas does this quite well. I'm not a UX/UI designer, so I'm hoping to get some help from the community on this one.

> and lacks governance (can I make the tag "cheese" and then tag it with "3rd-party-tags" with no approval?). 
Yup, agree with this. These are some issues that I list in "Unresolved questions" section. Would love to get your and the communities thought on this. 

I took your feedback about "too generic" into consideration when doing the latest update. I tried to angle this more as a "search term" thing, rather than the special cases I had listed before. But on the other hand, if a small team (without LinkedIn's resources) decides that they want to tag their entities with `pii`/`public` or `broken`/`good_for_use` instead of digging in to the source code and implementing a proper metadata-field (and the associated search and relationship builders) for each way they might consider using tags for, then I think the _enabling_ use case is quite clear.

File: docs/rfc/active/tags/README.md [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+- add `Tags` to aspects for entities.
+- Implement relationship builders as needed.
+- The implementation of and need for access control to tags is an open question
+- As this is first and foremost a tool for discovery, the UI work is extensible:
+  - Creating tags in a way that makes duplication and spelling mistakes difficult.
+  - Attaching tags to entities: autocomplete, dropdown, etc.
+  - Visualizing existing tags, and which are most popular?
+- Explore the concept of tagged tags for building tag hierarchies. What are the possible uses case (for example
+  "sensitivity" and "PII")? How is this visualized in the UI?
```
Comment (woodyzhou) [inline docs/rfc/active/tags/README.md:202]: I don't think "tag" (especially it's a tag on aspect level) is the best solution for regulation compliance. LinkedIn team has several mature solution and experience which we believe can share and benefit open source world.

File: docs/rfc/active/tags/README.md [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+We should create/update user guides to educate users for:
+
+- Suggestions on how to use tags: low threshold metadata-addition, and the possibility of doing new types of searches
+
+## Drawbacks
+
+This is definitely more complex than just adding strings to an array.
+
+## Alternatives
```
Comment (woodyzhou) [inline docs/rfc/active/tags/README.md:169]: I know it sounds a little bit weird to model and implement a tag system for metadata in a non-metadata model way. Fundamentally, a minimal tag solution requirements are: 1) link a string to an ID of something; 2) provide ID to tags and tag to IDs search. Have you considered to design and impl this tag solution as a built-in feature for a normal application? Will it be a even flexible design?

File: docs/rfc/active/tags/README.md (hunk 6 above)
Comment (frsann) [reply docs/rfc/active/tags/README.md:202]: Ok, this sensitivity solution is clearly a red herring, so I'll change it 😄

Comment (frsann) [reply docs/rfc/active/tags/README.md:202]: That being said, open sourcing your sensitivity implementation would be greatly appreciated

File: docs/rfc/active/tags/README.md (hunk 7 above)
Comment (frsann) [reply docs/rfc/active/tags/README.md:169]: Sure, the ID-> string approach is one way, but you still need to "attach" it to objects, and I guess you'd want to have the ability to tag a object with more than one tag, so some sort of a "container" is needed. This is basically what we have right now. The only extra properties we're introducing are some AuditStamps and a descriptions (@jplaisted  was worried that tags go undocumented). The permission-to-edit functionality is an open question, and can be removed if it adds too much complexity. But I can mention this view in this section. Thanks!

>  built-in feature for a normal application? Will it be a even flexible design?

Sorry, I'm not sure what you mean by this? Can you elaborate, please?

Comment (woodyzhou) [reply docs/rfc/active/tags/README.md:169]: please ignore my below comment, it's implementation details.

>built-in feature for a normal application? Will it be a even flexible design?

File: docs/rfc/active/tags/README.md (hunk 4 above)
Comment (jplaisted) [reply docs/rfc/active/tags/README.md]: So I think at this point the two of us may have to agree to disagree. Let's rope in more people :) @shirshanka @mars-lan @pardhugunnam 

Final thoughts below

> This is a UI problem in my opinion. 

Again, maybe working on use cases and proposed UI features first would help figure out what we want here. Favorites? GDPR taxonomy? 3rd party access rights? To me, these are all different use cases. Put another way, maybe top down would work better than bottom up here :)

I'll also say that typos can come from outside the UI. Imagine I'm a developer writing a crawler now, and I want the crawler to "tag" the metadata it is finding. If it is just a generic string field, I can typo values in code. Or if I'm writing a query, I can typo that too. Note that typo doesn't even need to be spelling something the wrong way; case sensitivity often also matters. But if we're defining a well known taxonomy (this is the word I was looking for over lexicon earlier), and we have a set of well known constants we can use in code, and then if users need to use in a UI then we can use a drop down from said taxonomy, we're in a better state, imo. It also has the advantage of being quite a bit clearer, imo. Even if we just define string constants, like `pii`, if all we have is an undefined set of tags (which can include these constants, but does not have to), there's no clear connection between the two. But if have some `Compliance` structure that has some `hasPii` field, it is quite a bit clearer imo. I can search the code for `pii` and figure out where I'm meant to read/write it, where as with just `tags` it is unclear.

> I'm not a UX/UI designer

Neither am I. I'm a backend engineer trying to wear a product hat here, admittedly. 

> The main benefit I see in this is the "enabling" possibilities: a small team (without LinkedIn's resources and scale) that decides that they want to tag their entities with pii/public

- Good point that maybe we need more focus on smaller teams.
- However, if we add this feature, we're adding it for everyone, big and small.
- I think if we do add this, it is kind of a pandora's box. Once open it is hard to close, and its hard to ask people to do the right thing (e.g. "don't add custom GDPR taxonomy via tags"). Put another way, I do think there's value in "saving people from themselves" and being opinionated on how to do things.
- Again, to me, the `pii` and `public` examples are no-goes here. We should not encourage that. Those are actionable tags. Especially from a legal standpoint.
- If the goal here is to make it easier to get different kinds of metadata into DataHub (which can include metadata classifications), maybe we should focus on making that easier, rather than making a very generic and potentially dangerous way to do it? I agree that adding metadata models, search models, an index builder, add restli end points, and editing the UI, is quite a bit of work and potentially overly complicated at the moment. It'll probably be a lot of work to make that process smoother, but that might be a better end state. Make it much easier to customize DataHub's stack, from models to UI, all the changes of which can be reviewed (governance) and you end up with well defined models/taxonomies rather than just a set of undefined strings with tags. So if an organization wants their own GDPR taxonomy, they can define it (again, we have one at LI, but it isn't open source, idk why; but imagine it was but someone else wanted their own definition). Or if you want to model "favorite" datasets, consider adding that as a first class model (which, again, we have at LI). I do think there's a lot of power in actually having well defined models over undefined sets of strings.
   - Ideas to improve:
      - I don't think there's any real reason why most people have to define restli endpoints / GMSes themselves. Most of it won't be customized, but just writing up Elasticsearch/Neo4j and then inheriting from our base restli resource class with no changes in implementation. I think it is reasonable that we could just auto generate this from a small config file.
      - Changing models requires forking DataHub right now, this isn't great. It'd be nice if we decoupled the models and DataHub repo. I think this requires our next generation metadata events though (right now, "MXE v4" needs to reference every single model transitively).
      - I'm not sure how easy it is to edit the DataHub UI. cc @jjoyce0510 as an fyi for the new react UI - might want to make it "easy" to be able to edit the UI with your custom models.
   

tl;dr to me, the "right thing to do" is to have well defined models for your metadata. If that process is difficult right now, we should focus on how to improve that, rather than open a potentially dangerous side channel.

Comment (shirshanka) [reply docs/rfc/active/tags/README.md]: I looked briefly at a few other mature enterprise catalog systems that do this. 

Apache Atlas : 
- Seems to have tags: https://docs.cloudera.com/HDPDocuments/HDP2/HDP-2.5.3/bk_data-governance/content/ch_working_with_atlas_tags.html
- Seems to also have glossary / taxonomy: https://docs.cloudera.com/HDPDocuments/HDP2/HDP-2.5.3/bk_data-governance/content/ch_managing_atlas_business_taxonomy.html

Collibra: 
- Documentation is scarce but it seems like they support user-defined tags as well as glossary
https://www.collibra.com/newsroom/collibra-5-1-enables-new-levels-collaboration-productivity-data-governance-business-users
> Tagging: Further strengthening the Collibra approach to crowd-sourced metadata, Collibra 5.1 will feature on-the-fly tagging capabilities that leverage the power of large numbers of data citizens to enrich the information available for data sets. Tags can be filtered and stored in a view definition to make data collections easier to find.
- https://university.collibra.com/knowledge/collibra-body-of-knowledge/import-content-from-excel/importing-business-terms-from-excel-spreadsheets/basics-of-importing-business-terms/


Alation:
- Has a REST API to work with tags, seems simple: https://customerportal.alationdata.com/docs/TagsAPI/index.html
- Couldn't find anything about taxonomies

Side rant: how come enterprise catalogs have such bad documentation!!!

It seems to me that glossaries are being targeted mainly for semantic business understanding to describe data in business terms (e.g. Customer, Financial Statement) versus more technical understanding (e.g. pii). However, I can see it being done both ways. The way LinkedIn did it is more akin to a "fixed taxonomy", but we have faced some challenges in figuring out how we evolve from it. 

Given that these are pretty mature catalog systems that have spent time understanding customer needs and have evolved in reaction to that, it seems clear that there is some need for user-defined and flexible tags as well as platform-team or governance-team controlled taxonomies or business glossaries. 

We can probably let individual companies decide where their tolerance for "flexible tag graph" versus "strict hierarchical taxonomy" lies, and let them decide whether their "pii" annotations live in the tags cloud (as proposed by this rfc) or inside a strict taxonomy (an example of which is the business glossary rfc).

File: docs/rfc/active/tags/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (jjoyce0510) [reply docs/rfc/active/tags/README.md:55]: The benefit of modeling as a top level entity type is that the same tag is defined once, thus helping achieve semantic consistency. We can support autocomplete on tags, as well as the addition of metadata like description in a single place. 

Think of tags like "hashtags". I should be able to click a tag and see all entities associated with the tag.

**Comment by shirshanka** [comment]
Getting back to this, and after discussion with @woodyzhou; I'm comfortable with supporting this as a simple global tags as a carrier for strings, to support search by tag and user tagging an entity scenarios. 

Non-goals for this rfc: 
Access control on tags -> will be tackled through access control on metadata model as a separate problem. 
Support for taxonomies -> will be tackled through the business glossary work.
Support for tags on tags -> slippery slope towards taxonomy

File: docs/rfc/active/tags/README.md [hunk 8]
```diff
@@ -0,0 +1,190 @@
+- Start Date: 2021-02-17
+- RFC PR: (after opening the RFC PR, update this with a link to it and update the file name)
```
Comment (shirshanka) [inline docs/rfc/active/tags/README.md:2]: Could you add the PR number here?: 2112

Comment (frsann) [reply docs/rfc/active/tags/README.md]: Done

**Comment by shirshanka** [comment]
@hsheth2 : can you help @frsann figure out why the docs build is broken and how to fix it? 
Presumably: add the rfc doc to the sidebar? 
And does that mean we need to add to the "How to create an RFC" doc? Sounds a bit onerous.

**Comment by hsheth2** [comment]
@frsann you’ll have to add an entry to `docs-website/sidebars.js`. I’m working on a better way to automate it so that nothing extra is required - should be out soon.

**Comment by frsann** [comment]
> @frsann you’ll have to add an entry to `docs-website/sidebars.js`. I’m working on a better way to automate it so that nothing extra is required - should be out soon.

Thanks! I rebased and updated `sidebars.js`.

**Review by shirshanka — APPROVED** [review]
Merging this in. 
We will continue learning and build the right guardrails through future rfc-s, as the community tries the feature out, to balance between free-form tags and taxonomies.
